src/vanilla/utils/atomWithStorage.ts

```diff
@@ -114,25 +116,40 @@ const defaultStorage = createJSONStorage(() =>
     : (undefined as unknown as Storage)
 )
 
+export function atomWithStorage<Value>(
+  key: string,
+  initialValue: Value,
+  storage?: AsyncStorage<Value>
+): WritableAtom<
+  Promise<Value> | Value,
+  [SetStateActionWithReset<Promise<Value> | Value>],
+  Promise<void>
+>
+
+export function atomWithStorage<Value>(
+  key: string,
+  initialValue: Value,
+  storage?: SyncStorage<Value>
+): WritableAtom<Value, [SetStateActionWithReset<Value>], void>
+
 export function atomWithStorage<Value>(
   key: string,
   initialValue: Value,
   storage:
     | SyncStorage<Value>
     | AsyncStorage<Value> = defaultStorage as SyncStorage<Value>
-): WritableAtom<Value, [SetStateActionWithReset<Value>], void> {
-  const baseAtom = atom(initialValue)
+): any {
```

Comment:
`any` is unfortunate here, but can't be helped as we overload functions and TS can't understand it.